playwright/src/main/java/com/microsoft/playwright/junit/impl/PlaywrightExtension.java

```diff
@@ -4,18 +4,50 @@
 import com.microsoft.playwright.junit.Options;
 import org.junit.jupiter.api.extension.*;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
 import static com.microsoft.playwright.junit.impl.ExtensionUtils.isParameterSupported;
 import static com.microsoft.playwright.junit.impl.ExtensionUtils.setTestIdAttribute;
 
-public class PlaywrightExtension implements ParameterResolver, AfterAllCallback {
-  private static final ThreadLocal<Playwright> threadLocalPlaywright = new ThreadLocal<>();
+public class PlaywrightExtension implements ParameterResolver, BeforeAllCallback, ExtensionContext.Store.CloseableResource {
+  private static final ThreadLocal<Playwright> threadLocalPlaywright;
+  private static final List<Playwright> playwrightList;
+  private final static Lock beforeLock;
+  private static boolean isTestRunStarted;
+
+  static {
+    isTestRunStarted = false;
+    beforeLock = new ReentrantLock();
+    threadLocalPlaywright = new ThreadLocal<>();
+    playwrightList = new ArrayList<>();
+  }
 
+  // Before a Test class starts, we register a closeable resource
+  // We use a lock + boolean to ensure this only gets run once for the entire test run
   @Override
-  public void afterAll(ExtensionContext extensionContext) {
-    Playwright playwright = threadLocalPlaywright.get();
-    threadLocalPlaywright.remove();
-    if (playwright != null) {
-      playwright.close();
+  public void beforeAll(final ExtensionContext context) {
+    try {
+      beforeLock.lock();
+      if (!isTestRunStarted) {
+        isTestRunStarted = true;
+        context.getRoot().getStore(ExtensionContext.Namespace.GLOBAL).put(context.getUniqueId(), this);
+      }
+    } finally {
+      beforeLock.unlock();
+    }
+  }
+
+  // This is a workaround for JUnit's lack of an "AfterTestRun" hook
+  // This will be called once after all tests have completed.
+  @Override
+  public void close() throws Throwable {
+    for (Playwright playwright : playwrightList) {
```

Comment:
This way we close playwright instance that is still referenced from a thread local. To avoid that we could have a pool of playwright instances, so that whenever a test on a thread starts, it borrows a spare playwright from the pool or creates a new one. After all test methods have finished it must return the playwright instance to the pool and clear thread local reference to it. If next test starts on the same thread, it will borrow playwright from the pool again. This assumes that all tests within the class run on the same thread (which I believe is our assumption in other places too. After all tests have completed there should be no thread local references to the playwright instances and we can safely  close all of them. What do you think?

Reply:
> This way we close playwright instance that is still referenced from a thread local. 

my understanding is that ThreadLocals are closed when the thread is killed so do we need to worry about this?  This close method will only run after all tests have run so all threads are done at that point and no future threads will launch.  Is your concern that closing playwright here will fail?  I tested this and didn't see any exceptions being thrown.  Apologies for being dense here but i don't understand the concern.

I'm not opposed to doing some sort pool for it but I think it will take more work.  Maybe we can do it in a future iteration?

Reply:
> my understanding is that ThreadLocals are closed when the thread is killed so do we need to worry about this?

Yes, you are right. Java docs state that [explicitly](https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html) " after a thread goes away, all of its copies of thread-local instances are subject to garbage collection (unless other references to these copies exist)." Nothing to worry about then, we can collect them in a list and close after all tests are done. There should be no usages after that point, unless the tests posted a reference that will be accessed on a different thread, so it should be safe to close for us.